packages/view/lib/view.js

```diff
@@ -0,0 +1,383 @@
+
+        if (
+          !every(values || {}, function(value, path) {
+            return value === true && path in ctx ? true : ctx[path] === value;
+          })
+        ) {
+          return this;
+        }
+      }
+
+      this.actionQueue.push(callback);
+    } else if (on === 'init') {
+      /* Init events are always fired in series, before any other actions
+       *
+       * Example:
+       *     view.on('init', function (next) {
+       *         // do something before any actions or queries have run
+       *     });
+       */
+
+      this.initQueue.push(callback);
+    } else if (on === 'render') {
+      /* Render events are always fired last in parallel, after any other actions
+       *
+       * Example:
+       *     view.on('render', function (next) {
+       *         // do something after init, action and query middleware has run
+       *     });
+       */
+
+      this.renderQueue.push(callback);
+    }
+
+    // TODO: Should throw if we didn't recognise the first argument!
+
+    return this;
+  }
+
+  /**
+   * Queues a mongoose query for execution before the view is rendered.
+   * The results of the query are set in `locals[key]`.
+   *
+   * Keys can be nested paths, containing objects will be created as required.
+   *
+   * The third argument `then` can be a method to call after the query is completed
+   * like function(err, results, callback), or a `populatedRelated` definition
+   * (string or array).
+   *
+   * Examples:
+   *
+   * view.query('books', keystone.list('Book').model.find());
+   *
+   *     an array of books from the database will be added to locals.books. You can
+   *     also nest properties on the locals variable.
+   *
+   * view.query(
+   *     'admin.books',
+   *      keystone.list('Book').model.find().where('user', 'Admin')
+   * );
+   *
+   *     locals.admin.books will be the result of the query
+   *     views.query().then is always called if it is available
+   *
+   * view.query('books', keystone.list('Book').model.find())
+   *     .then(function (err, results, next) {
+   *         if (err) return next(err);
+   *         console.log(results);
+   *         next();
+   *     });
+   *
+   * @api public
+   */
+
+  query(key, listOrKey, query, options) {
+    let list = listOrKey;
+    if (typeof listOrKey === 'string') {
+      list = this.keystone.getListByKey(listOrKey);
+    }
+    let locals = this.res.locals;
+    const parts = key.split('.');
+    const chain = new QueryCallbacks(options);
+
+    key = parts.pop();
+
+    for (let i = 0; i < parts.length; i++) {
+      if (!locals[parts[i]]) {
+        locals[parts[i]] = {};
+      }
+      locals = locals[parts[i]];
+    }
+
+    this.queryQueue.push(async () => {
+      const callbacks = chain.callbacks;
+      const context = this.keystone.getAccessContext('admin', this.req);
+      try {
+        const results = await list.listQuery(query, context, list.gqlNames.listQueryName);
```

Comment:
@jesstelford @JedWatson: is `list.listQuery` right way to call this from within server side code? Can I use mutations similarly (`createMutation`, `updateMutation` and so on)?

I could not find any better way to do this. Still feel there can be improvement. In gql query you must specify the fields you want to retrieve. How do I do it here? filter from result? Let me know if there is better way to use the gql layer properly, I would also want to use the gql layer with `skipAccessControl` as my server side logic may already handle more complex scenario sometime.

Reply:
I think the most useful thing to expose here would be the apollo server query API. That way you can get the full power of our graphQL queries and don't have to reinvent wheels like selecting fields, filtering results, etc.

```javascript
      const schemaName = 'admin';
      const queryFn = this.keystone._graphQLQuery[schemaName];
      const context = this.keystone.getAccessContext(schemaName, this.req);
      try {
        const { data, errors } = queryFn(query, context);
        ....
```
if you want to turn off access control, you can do a similar thing to what we do in hooks, which is:

```javascript
        if (skipAccessControl) {
          passThroughContext = {
            ...context,
            getListAccessControlForUser: () => true,
            getFieldAccessControlForUser: () => true,
          };
        }
        ...
        const { data, errors } = queryFn(query, passThroughContext);
```

At the moment `_graphQLQuery` isn't technically part of the public API, but we might want to change this in the future. For now feel free to work as if it were.

Reply:
thanks, this greatly helps. 

updated.